src/coreclr/tools/Common/Compiler/TypeExtensions.cs

```diff
@@ -682,6 +682,28 @@ private static TypeDesc GetTypeThatMeetsConstraints(GenericParameterDesc generic
                 if (c.ContainsSignatureVariables())
                     return null;
 
+                // If there's unimplemented static abstract methods, this is not a suitable instantiation.
```

Comment:
It will use __Canon (or even __UniversalCanon if universal shared code is available, which it isn't) if `allowCanon` is true, but we can't currently use __Canon everywhere:

https://github.com/dotnet/runtime/blob/5369b096c4675df82c923e5fd0d278465f8feb79/src/coreclr/tools/Common/Compiler/TypeExtensions.cs#L642

There's a file format limitation around generic methods (and the bug happens in a mal-instantiated generic method):

https://github.com/dotnet/runtime/blob/8e8a62156ea459945b1d953c4fbe950d276bbc9b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/RootingHelpers.cs#L67-L70